### PR TITLE
Use crypto nonces from proto messages in C++ and Java

### PR DESCRIPTION
**Pull request by ipetr0v**
This PR makes C++ and Java crypto code to use nonces provided by Proto messages.

Ref https://github.com/project-oak/oak/issues/4507

### CODE REVIEW

File: cc/crypto/client_encryptor.cc [hunk 1]
```diff
@@ -69,8 +70,10 @@ absl::StatusOr<EncryptedRequest> ClientEncryptor::Encrypt(absl::string_view plai
 
 absl::StatusOr<DecryptionResult> ClientEncryptor::Decrypt(EncryptedResponse encrypted_response) {
   // Decrypt response.
+  const std::vector<uint8_t> nonce(encrypted_response.encrypted_message().nonce().begin(),
```
Comment (conradgrobler) [inline cc/crypto/client_encryptor.cc:73]: Is this copy required? Can't you just pass in encrypted_response.encrypted_message().nonce()?

File: cc/crypto/hpke/recipient_context.cc [hunk 2]
```diff
@@ -118,14 +118,14 @@ std::vector<uint8_t> RecipientContext::GenerateNonce() {
   return nonce;
 }
 
-absl::StatusOr<std::string> RecipientContext::Open(absl::string_view ciphertext,
+absl::StatusOr<std::string> RecipientContext::Open(const std::vector<uint8_t>& nonce,
```
Comment (conradgrobler) [inline cc/crypto/hpke/recipient_context.cc:121]: Related to the previous point: why is this a vector reference rather than a stringview?

**Review by conradgrobler — APPROVED** [review]

File: cc/crypto/client_encryptor.cc (hunk 1 above)
Comment (ipetr0v) [reply cc/crypto/client_encryptor.cc:73]: This copy will happen anyway, because the BoringSSL decryption function requires a vector

Comment (conradgrobler) [reply cc/crypto/client_encryptor.cc:73]: It seem like it takes a pointer and a length: https://commondatastorage.googleapis.com/chromium-boringssl-docs/aead.h.html#EVP_AEAD_CTX_open. Our helper function takes a vector, but it seems like it doesn't have to: https://github.com/project-oak/oak/blob/78cba93869de13e6959c4b138f4e0fe9b1a6b3c3/cc/crypto/hpke/utils.cc#L130

**Review by tiziano88 — APPROVED** [review]

File: cc/crypto/client_encryptor.cc (hunk 1 above)
Comment (ipetr0v) [reply cc/crypto/client_encryptor.cc:73]: Will be done as part of https://github.com/project-oak/oak/issues/4600

**Pull request merged**